feat(native): add hardware identity support

chrysn opened the pull request:
# Description

This implements the `has_device_identity` / `hw/identity` feature for native based on the [standard `/etc/machine-id`](https://www.freedesktop.org/software/systemd/man/latest/machine-id.html) file.

As by the standard, we don't use the machine ID 1:1 (as that'd allow third parties to correlate our use of the ID with another application's use), but hash it with an application specific identifier.

As by the docs of ariel-os-identity (which BTW I think we should have built somewhere if only to link from here), this identifier stays constant across using different applications -- neither the application name nor the build path are factored in. This can be a bit weird because there *can* be multiple instances of a native application running at the same time, but then, if those are networked, the user will have to do manual assignment of the devices to tap interfaces etc as well.

This PR does not introduce any means for the user to override the machine ID (just as there is currently no override for the RNG in the form of a seed). I envision that at some point per-running-program configuration will be needed (most importantly, assigning a tap device), either in the form of an environment variable or a command line argument, and possibly through a config file, but there is no urgent need to do this here, so let's do this properly when we need it.

## Change checklist

- [x] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](https://ariel-os.github.io/ariel-os/dev/docs/book/coding-conventions.html).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
  *will the feature table update automatically? or scream in CI?*

ROMemories commented on src/ariel-os-native/Cargo.toml:
```diff
@@ -23,6 +23,7 @@ ariel-os-random = { workspace = true, optional = true }
 rand = { workspace = true, default-features = false, optional = true, features = [
   "getrandom",
 ] }
+sha2 = { version = "0.10.8", default-features = false }
```
Any reason not to use `0.10.9`?

chrysn replied:
Yes, I didn't want to have too much churn on the Cargo.toml, that just causes merge conflicts.

(The dependency should generally reflect what is intended and tested; I did test with 0.10.8, so it's fine, although I'm pretty sure I don't depend on anything beyond 0.10.0)

ROMemories replied:
> I didn't want to have too much churn on the Cargo.toml

I think you mean the `Cargo.lock`? We don't currently have a direct dependency on `sha2`.

Otherwise makes sense, even though that dependency could have been defined in the workspace's manifest now that I think about it, it's likely not going to be the only crate using it.

ROMemories commented on src/ariel-os-native/src/identity.rs:
```diff
@@ -0,0 +1,33 @@
+//! Provides a Device ID based on [machine-id]
```
```suggestion
//! Provides a Device ID based on [machine-id].
```

ROMemories commented on src/ariel-os-native/src/identity.rs:
```diff
@@ -0,0 +1,33 @@
+
+pub struct DeviceId([u8; SHA256LEN]);
+
+extern crate std;
```
That should be moved as the first item in this file (above the `use`s), similarly to `mod` declarations.

kaspar030 commented:
> As by the docs of ariel-os-identity (which BTW I think we should have built somewhere if only to link from here), this identifier stays constant across using different applications -- neither the application name nor the build path are factored in. 

ok

> This can be a bit weird because there _can_ be multiple instances of a native application running at the same time, but then, if those are networked, the user will have to do manual assignment of the devices to tap interfaces etc as well.

So I guess once there is networking, we'll need to maybe hash the interface path into this. The device identity is used to calculate the tap's Ariel side MAC address, so it needs to be unique.

chrysn commented on src/ariel-os-native/src/identity.rs:
```diff
@@ -0,0 +1,33 @@
+
+pub struct DeviceId([u8; SHA256LEN]);
+
+extern crate std;
```
Yes, making it std would be the right thing, will change this here.

ROMemories commented:
> There were a few weirdly non-interactable comments in the PR view. I tried whether a rebase-force-push helps GitHub make up its mind about them, but apparently it doesn't. @ROMemories, can you mark them as resolved?

To me all the threads show up as resolved by show, so I have nothing to do.

Why did you remove the `ci-build:small` label, is this not pretty much ready to merge?

chrysn commented:
I removed it temporarily while I was accepting changes, then sending my own changes addressing some comments, and didn't want to cancel to many CI jobs keeping others up.

Re-enabled now, and ready from my PoV. (Nothing to do on the documentation side, as native is not yet in the hardware matrix).
